### ISSUE TITLE
k8s module does not create namespaces/projects due to "cannot get namespaces in the namespace ..."

### ISSUE THREAD

**Issue by athimm** [ansible/ansible#58673]
##### SUMMARY
On an OpenShift 3.11 cluster a user is allowed to list and create projects via the oc CLI, e.g.
oc projects
oc new-project testing

We are trying to map this to k8s, with the following task:
```yaml
   - name: Create a k8s namespace
      k8s:
        api_version: v1
        kind: Namespace
        name: testing
        namespace: testing
        state: present
```
(also tried w/o "namespace: testing")
This fails with
```paste
    "msg": "Failed to retrieve requested object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"namespaces \\\\\"testing\\\\\" is forbidden: User \\\\\"XXX@YYY\\\\\" cannot get namespaces in the namespace \\\\\"testing\\\\\": no RBAC policy matched\",\"reason\":\"Forbidden\",\"details\":{\"name\":\"testing\",\"kind\":\"namespaces\"},\"code\":403}\\n'",
    "reason": "Forbidden",
    "status": 403
}
```

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
k8s

##### ANSIBLE VERSION
```
ansible 2.8.1
  config file = /home/thimm/.ansible.cfg
  configured module search path = ['/home/thimm/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.3 (default, May 11 2019, 00:38:04) [GCC 9.1.1 20190503 (Red Hat 9.1.1-1)]
```

##### CONFIGURATION
```
INTERPRETER_PYTHON(/home/thimm/.ansible.cfg) = auto
```

##### OS / ENVIRONMENT
Fedora 30 w/ default ansible rpm

##### STEPS TO REPRODUCE
This playbook was used for the tests
```yaml
--
- hosts: localhost
  gather_facts: no
  tasks:
    - name: Create a k8s namespace
      k8s:
        api_version: v1
        kind: Namespace
        name: testing
        namespace: testing
        state: present
```
##### EXPECTED RESULTS
A project/namespace should had been created

##### ACTUAL RESULTS
```paste
TASK [Create a k8s namespace] *****************************************************************************************************************************
task path: /storage3/home/thimm/SDX/site.yml:5
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: thimm
<127.0.0.1> EXEC /bin/sh -c 'echo ~thimm && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/thimm/.ansible/tmp/ansible-tmp-1562164734.5078351-37883863576367 `" && echo ansible-tmp-1562164734.5078351-37883863576367="` echo /home/thimm/.ansible/tmp/ansible-tmp-1562164734.5078351-37883863576367 `" ) && sleep 0'
Using module file /usr/lib/python3.7/site-packages/ansible/modules/clustering/k8s/k8s.py
<127.0.0.1> PUT /home/thimm/.ansible/tmp/ansible-local-694264lm_2jn/tmpstdf2zcg TO /home/thimm/.ansible/tmp/ansible-tmp-1562164734.5078351-37883863576367/AnsiballZ_k8s.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/thimm/.ansible/tmp/ansible-tmp-1562164734.5078351-37883863576367/ /home/thimm/.ansible/tmp/ansible-tmp-1562164734.5078351-37883863576367/AnsiballZ_k8s.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3 /home/thimm/.ansible/tmp/ansible-tmp-1562164734.5078351-37883863576367/AnsiballZ_k8s.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/thimm/.ansible/tmp/ansible-tmp-1562164734.5078351-37883863576367/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/tmp/ansible_k8s_payload_8sum8t4r/ansible_k8s_payload.zip/ansible/module_utils/k8s/raw.py", line 235, in perform_action
    existing = resource.get(**params)
  File "/usr/lib/python3.7/site-packages/openshift/dynamic/client.py", line 75, in inner
    raise api_exception(e)

fatal: [localhost]: FAILED! => {
    "changed": false,
    "error": 403,
    "invocation": {
        "module_args": {
            "api_key": null,
            "api_version": "v1",
            "append_hash": false,
            "ca_cert": null,
            "client_cert": null,
            "client_key": null,
            "context": null,
            "force": false,
            "host": null,
            "kind": "Namespace",
            "kubeconfig": null,
            "merge_type": null,
            "name": "testing",
            "namespace": "testing",
            "password": null,
            "resource_definition": null,
            "src": null,
            "state": "present",
            "username": null,
            "validate": null,
            "validate_certs": null,
            "wait": false,
            "wait_condition": null,
            "wait_timeout": 120
        }
    },
    "msg": "Failed to retrieve requested object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"namespaces \\\\\"testing\\\\\" is forbidden: User \\\\\"axel.thimm@sixdx.com\\\\\" cannot get namespaces in the namespace \\\\\"testing\\\\\": no RBAC policy matched\",\"reason\":\"Forbidden\",\"details\":{\"name\":\"testing\",\"kind\":\"namespaces\"},\"code\":403}\\n'",
    "reason": "Forbidden",
    "status": 403
}

PLAY RECAP ************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/clustering/k8s/k8s.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/clustering/k8s/k8s.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @chouseknecht @fabianvf @flaper87 @maxamillion @willthames
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by fabianvf** [comment]
You shouldn't supply the `namespace` parameter unless you are created namespaced resources, if you remove the `namespace` parameter it should work fine. We should probably throw a more clear error before we attempt to make the request though, since we can know it won't work .

**Comment by axelthimm** [comment]
Thanks for the fast reply! Originally I hadn't the namespace parameter, both variants fail in the same way. Here is a minimal playbook that fails w/o the namespace set:
```yaml
---
- hosts: localhost
  gather_facts: no
  tasks:
    - name: Create a k8s namespace
      k8s:
        api_version: v1
        kind: Namespace
        name: testing
```
(contents of site2.yml)
```paste
$ ansible-playbook -i ./inventory.yml ./site2.yml -vvv
ansible-playbook 2.8.1
  config file = /home/thimm/.ansible.cfg
  configured module search path = ['/home/thimm/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 3.7.3 (default, May 11 2019, 00:38:04) [GCC 9.1.1 20190503 (Red Hat 9.1.1-1)]
Using /home/thimm/.ansible.cfg as config file
host_list declined parsing /storage3/home/thimm/YYY/inventory.yml as it did not pass it's verify_file() method
script declined parsing /storage3/home/thimm/YYY/inventory.yml as it did not pass it's verify_file() method
Parsed /storage3/home/thimm/YYY/inventory.yml inventory source with yaml plugin

PLAYBOOK: site2.yml ***************************************************************************************************************************************
1 plays in ./site2.yml

PLAY [localhost] ******************************************************************************************************************************************
META: ran handlers

TASK [Create a k8s namespace] *****************************************************************************************************************************
task path: /storage3/home/thimm/YYY/site2.yml:5
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: thimm
<127.0.0.1> EXEC /bin/sh -c 'echo ~thimm && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/thimm/.ansible/tmp/ansible-tmp-1562173259.95595-207986756939115 `" && echo ansible-tmp-1562173259.95595-207986756939115="` echo /home/thimm/.ansible/tmp/ansible-tmp-1562173259.95595-207986756939115 `" ) && sleep 0'
Using module file /usr/lib/python3.7/site-packages/ansible/modules/clustering/k8s/k8s.py
<127.0.0.1> PUT /home/thimm/.ansible/tmp/ansible-local-12043wfrcdr7u/tmph7m6gpzi TO /home/thimm/.ansible/tmp/ansible-tmp-1562173259.95595-207986756939115/AnsiballZ_k8s.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/thimm/.ansible/tmp/ansible-tmp-1562173259.95595-207986756939115/ /home/thimm/.ansible/tmp/ansible-tmp-1562173259.95595-207986756939115/AnsiballZ_k8s.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3 /home/thimm/.ansible/tmp/ansible-tmp-1562173259.95595-207986756939115/AnsiballZ_k8s.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/thimm/.ansible/tmp/ansible-tmp-1562173259.95595-207986756939115/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/tmp/ansible_k8s_payload_i51p4331/ansible_k8s_payload.zip/ansible/module_utils/k8s/raw.py", line 235, in perform_action
    existing = resource.get(**params)
  File "/usr/lib/python3.7/site-packages/openshift/dynamic/client.py", line 75, in inner
    raise api_exception(e)

fatal: [localhost]: FAILED! => {
    "changed": false,
    "error": 403,
    "invocation": {
        "module_args": {
            "api_key": null,
            "api_version": "v1",
            "append_hash": false,
            "ca_cert": null,
            "client_cert": null,
            "client_key": null,
            "context": null,
            "force": false,
            "host": null,
            "kind": "Namespace",
            "kubeconfig": null,
            "merge_type": null,
            "name": "testing",
            "namespace": null,
            "password": null,
            "resource_definition": null,
            "src": null,
            "state": "present",
            "username": null,
            "validate": null,
            "validate_certs": null,
            "wait": false,
            "wait_condition": null,
            "wait_timeout": 120
        }
    },
    "msg": "Failed to retrieve requested object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"namespaces \\\\\"testing\\\\\" is forbidden: User \\\\\"XXX@YYY\\\\\" cannot get namespaces in the namespace \\\\\"testing\\\\\": no RBAC policy matched\",\"reason\":\"Forbidden\",\"details\":{\"name\":\"testing\",\"kind\":\"namespaces\"},\"code\":403}\\n'",
    "reason": "Forbidden",
    "status": 403
}

PLAY RECAP ************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

**Comment by willthames** [comment]
@axelthimm that's an RBAC issue, not an ansible issue 

`"message":"namespaces \"testing\" is forbidden: User \"XXX@YYY\" cannot get namespaces in the namespace \"testing\": no RBAC policy matched"`

**Comment by willthames** [comment]
The k8s module will check whether a namespace exists before creating one, so it needs `get` permission as well as `create`

**Comment by axelthimm** [comment]
@willthames, I also suspected that k8s creation process involves querying first then creating. I'm not sure that is the full picture as "oc projects" that lists the namespaces does work. Probably the query methods of k8s and "oc projects" must differ in some significant way to have this distinction.

Although I'm not a fan of the RBAC model used in the specific case, it does work for oc CLI, helm charts, go scripts etc, because they probably do not query first, or query differently. Which is why I would still consider this an ansible/module issue as we would much prefer to be able to map this onto ansible.

ATM I'm using a command module workaround that calls "oc new-project" and then k8s seems to be pretty much able to do all the rest, but it would be nicer to have a pure k8s solution. Maybe a flag that ask of k8s to perform creates blindly and check for the API response whether the entity was there or not, or maybe always create w/o a get operation. I guess this is what "oc new-project" does.

In case this is not possible due to k8s' design to query first in a specific way and then create/update, would it make sense to document this requirement and especially the rationale behind it? The latter would make it much easier to argue with openshift/kubernetes admins that have setup a too stringent RBAC model.

**Comment by willthames** [comment]
Actually there should be no need to check for namespaces in a namespace - that's nonsense, so we might be doing the wrong thing.

**Comment by willthames** [comment]
Ok, that is just the standard error if you don't have `get` permissions on `namespaces`:

```
$ kubectl --kubeconfig ~/.kube/oo.config get ns kube-public
Error from server (Forbidden): namespaces "kube-public" is forbidden: User "system:serviceaccount:default:issue-58763" cannot get namespaces in the namespace "kube-public"
```

The k8s module relies on getting the existing thing before modifying for a number of reasons:
* It doesn't even try to create/update resources if the resource is in the desired state
* If it does change a resource, it returns the diff against the previously existing resource

Feel free to create patches against the documentation for improvements, but having `create` but not `get` permission on Namespaces is so weird that I'm doing no more on this problem than reviewing documentation improvements

**Comment by axelthimm** [comment]
I did some more digging and it turns out that I was comparing apples and oranges --- here namespaces and projects.
I was able to list projects and create projects with oc project/new-project and naively thought there are superentities of namespaces (which they kind of are, but then again not) - the RBAC does allow to query/create projects, but not namespaces. Quite an obscure setup, but that's the case here.
So changing the playbook's task to
```yaml
    - name: Create a k8s project
      k8s:
        api_version: v1
        kind: Project
        name: testing
```
everything falls in line (note the _kind: **Project**_).
Sorry for barking up the wrong tree, this is a layer 8 problem. At least the issue archive may help someone in a similar position in the future.

Thanks for taking your time to look into this!
